internal/admission/validation/gateway/httproute.go

```diff
@@ -97,22 +98,69 @@ func validateHTTPRouteListener(listener *gatewayapi.Listener) error {
-				}
+	unsupportedHTTPRouteFilterTypes := []gatewayapi.HTTPRouteFilterType{
+		gatewayapi.HTTPRouteFilterURLRewrite,
+		gatewayapi.HTTPRouteFilterRequestMirror,
+	}
+	unsupportedFilterMap := lo.SliceToMap(unsupportedHTTPRouteFilterTypes, func(t gatewayapi.HTTPRouteFilterType) (gatewayapi.HTTPRouteFilterType, struct{}) {
+		return t, struct{}{}
+	})
+
+	const (
+		KindService = gatewayapi.Kind("Service")
+		KindGateway = gatewayapi.Kind("Gateway")
+	)
+
+	// Validate parentRefs: only allow empty, `/Gateway` or `gateway.networking.k8s.io/Gateway`
+	// REVIEW: Should this happen before we get here? we search for parent gateways before we reach here.
+	for parentRefIndex, parentRef := range httproute.Spec.ParentRefs {
+		if parentRef.Group != nil && *parentRef.Group != "" && *parentRef.Group != gatewayapi.V1Group {
+			return fmt.Errorf("parentRefs[%d]: %s is not a supported group for httproute parentRefs, only %s is supported",
+				parentRefIndex, *parentRef.Group, gatewayapi.V1Group)
+		}
+		if parentRef.Kind != nil && *parentRef.Kind != "" && *parentRef.Kind != KindGateway {
+			return fmt.Errorf("parentRefs[%d]: %s is not a supported kind for httproute parentRefs, only kind %s is supported",
+				parentRefIndex, *parentRef.Kind, KindGateway)
+		}
+	}
+
+	for ruleIndex, rule := range httproute.Spec.Rules {
+		// rule timeout is not supported.
+		if rule.Timeouts != nil {
+			return fmt.Errorf("rules[%d]: rule timeout is unsupported", ruleIndex)
+		}
+		// Filters URLRewrite, RequestMirror are not supported.
+		for filterIndex, filter := range rule.Filters {
+			if _, unsupported := unsupportedFilterMap[filter.Type]; unsupported {
+				return fmt.Errorf("rules[%d].filters[%d]: filter type %s is unsupported",
+					ruleIndex, filterIndex, filter.Type)
 			}
 		}
-		// We don't support any backendRef types except Kubernetes Services.
-		for _, ref := range rule.BackendRefs {
+
+		for refIndex, ref := range rule.BackendRefs {
+			// Specifying filters in backendRef is not supported.
+			if len(ref.Filters) != 0 {
+				return fmt.Errorf("rules[%d].backendRefs[%d]: filters in backendRef is unsupported",
+					ruleIndex, refIndex)
+			}
+			// We don't support any backendRef types except Kubernetes Services.
 			if ref.BackendRef.Group != nil && *ref.BackendRef.Group != "core" && *ref.BackendRef.Group != "" {
-				return fmt.Errorf("%s is not a supported group for httproute backendRefs, only core is supported", *ref.BackendRef.Group)
+				return fmt.Errorf("rules[%d].backendRefs[%d]: %s is not a supported group for httproute backendRefs, only core is supported",
+					ruleIndex, refIndex, *ref.BackendRef.Group)
+			}
+			if ref.BackendRef.Kind != nil && *ref.BackendRef.Kind != KindService {
+				return fmt.Errorf("rules[%d].backendRefs[%d]: %s is not a supported kind for httproute backendRefs, only %s is supported",
+					ruleIndex, refIndex, *ref.BackendRef.Kind, KindService)
```

Comment:
I'd group this condition so that if either group or kind or both are wrong, we display the same message only once. Something like "this <group/kind> is not supported, only core/service" is supported.

Reply:
Grouping the conditions brings extra check for `nil`s in `Group` or `Kind` fields in composing error messages because either one can be empty. If we do not include the input group/kind in error messages, it would be simpler and clearer to group the conditions. Otherwise(include group/kind in error messages), I would keep the code as is.